### DISCUSSION TITLE
AWS Authentication

### DISCUSSION THREAD

**Discussion by yesoreyeram** [grafana/grafana-infinity-datasource#501] [Specific APIs] [answered]
How to pull data from any AWS api using infinity

**Comment by yesoreyeram** [comment, marked as answer]
# Connecting AWS API in Grafana using Infinity

Support for connecting to AWS API is available from version 1.3.0

## Steps to connect to AWS APIs

1. Create a new service account in [AWS console](https://us-east-1.console.aws.amazon.com/iam/home#/users$new?step=details). ( AWS Console -> IAM -> Access Management -> Users -> Add users)
   1. Select **Access key - Programmatic access** as AWS Credentials type
   2. Set required permissions ( preferably CloudWatch ReadOnly Permission )
   3. Copy the access key and secret key
2. Install the infinity plugin in Grafana and add data source for the same
3. Expand Authentication section and select "AWS"
4. Select region. Example `us-east-1` 
5. Select service. Example `monitoring`. You can find the appropriate service name either [here](https://github.com/aws/aws-sdk-go/blob/main/aws/endpoints/defaults.go#L224) or [here](https://docs.aws.amazon.com/general/latest/gr/aws-service-information.html).
6. Enter the access key and secret key you copied in step 1
7. Enter `https://monitoring.us-east-1.amazonaws.com` as allowed URL. (replace the service name and region as necessary )
8. Click "Save and Test"
9. Click the Explore button
10. Enter the URL `https://monitoring.us-east-1.amazonaws.com?Action=ListMetrics`
11. Select "JSON" as Query type
12. Optionally, Select "Backend" / "UQL" as parser.
13. Enter the Root Selector `ListMetricsResponse.ListMetricsResult.Metrics`
14. Click Run Query and you will see the results

## Config Editor

<img width="1339" alt="image" src="https://user-images.githubusercontent.com/153843/210791648-7d05d435-2a26-469c-9bfd-e4db98018999.png">


## Query with Backend parser 

<img width="1323" alt="image" src="https://user-images.githubusercontent.com/153843/210788954-e8bf3fab-e1c7-426d-8e87-610315c6afee.png">

## Query with UQL parser

<img width="1686" alt="image" src="https://user-images.githubusercontent.com/153843/210791302-178391c9-93f9-4449-8f5a-8e14a3db1eff.png">

sample uql query is given below

```kusto
parse-json
| scope "ListMetricsResponse.ListMetricsResult.Metrics"
| mv-expand "dimension"="Dimensions"
| project "Namespace", "MeasureName", "Dimension Name"="dimension.Name", "Dimension Value"="dimension.Value"
```

## Query with Default/Frontend parser

<img width="1791" alt="image" src="https://user-images.githubusercontent.com/153843/210790702-af822bdc-e974-4410-83b2-8e7776f03516.png">

**Comment by ramongalate** [comment]
is there anyway to do this with passwordless methods with roles and assume permissions?
safely adding credentials to systems like this is hard.

**Reply by elkh510** [reply]
https://github.com/grafana/grafana-infinity-datasource/pull/1538
